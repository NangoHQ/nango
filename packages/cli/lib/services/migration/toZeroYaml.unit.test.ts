import { describe, expect, it } from 'vitest';

import * as zeroYaml from './toZeroYaml.js';

import type { NangoModel } from '@nangohq/types';

describe('transformSync', () => {
    it('should transform a sync', () => {
        const ts = `import type { Model, NangoSync } from "../../models";
export default async function fetchData(nango: NangoSync) {
    await nango.log('hello');
    await nango.batchSave<Model>([{
        'id': 'foobar',
    }], 'Model');

}
export async function onWebhookPayloadReceived(
  nango: NangoSync,
  payload: any,
): Promise<void> {
  await nango.log('Received webhook', payload);
  await nango.batchSave<Model>([{}], 'Model');
}`;
        const result = zeroYaml.transformSync({
            content: ts,
            sync: {
                type: 'sync',
                name: 'test',
                description: 'Test sync',
                version: '1.2.3',
                runs: 'every hour',
                auto_start: true,
                sync_type: 'full',
                track_deletes: false,
                scopes: ['repo', 'user'],
                endpoints: [{ method: 'GET', path: 'top', group: 'foobar' }],
                usedModels: ['Metadata', 'Model', 'ModelDynamic'],
                webhookSubscriptions: ['*'],
                input: 'Metadata',
                output: ['Model', 'ModelDynamic']
            },
            models: new Map<string, NangoModel>([
                [
                    'Model',
                    {
                        name: 'Model',
                        fields: [
                            { name: 'id', value: 'string', tsType: true },
                            { name: 'undefined', value: 'undefined', tsType: true },
                            { name: 'arrTsType', value: 'string', tsType: true, array: true },
                            { name: 'null', value: 'null', tsType: true },
                            { name: 'bool', optional: false, tsType: true, value: 'boolean' },
                            { name: 'any', tsType: true, value: 'any' },
                            { name: 'opt', value: 'number', tsType: true, optional: true },
                            { name: 'ref', value: 'Foo', tsType: false, model: true, optional: true },
                            {
                                name: 'union',
                                union: true,
                                value: [
                                    { name: '0', value: 'literal1' },
                                    { name: '1', value: 'literal2' }
                                ],
                                tsType: false
                            },
                            {
                                name: 'array',
                                array: true,
                                value: [
                                    { name: '0', value: 'arr1' },
                                    { name: '1', value: 'arr2' }
                                ],
                                tsType: false
                            },
                            { name: 'obj', value: [{ name: 'nes', value: 'ted' }] },
                            {
                                name: 'dynamicObj',
                                optional: false,
                                value: [{ dynamic: true, name: '__string', optional: false, tsType: true, value: 'date' }]
                            }
                        ]
                    }
                ],
                [
                    'ModelDynamic',
                    {
                        name: 'ModelDynamic',
                        fields: [
                            { dynamic: true, name: '__string', optional: false, tsType: true, value: 'string' },
                            { name: 'id', optional: false, tsType: true, value: 'string' }
                        ]
                    }
                ],
                [
                    'Metadata',
                    {
                        name: 'Metadata',
                        fields: [
                            { name: 'foo', value: 'bar' },
                            { name: 'model', model: true, value: 'Model' }
                        ]
                    }
                ]
            ])
        });
        expect(result).toMatchSnapshot();
    });
});

describe('transformAction', () => {
    it('should transform an action', () => {
        const ts = `import type { NangoSync, IssueOutput, IssueInput } from "../../models";

export default async function runAction(nango: NangoSync, input: IssueInput): Promise<IssueOutput> {
  await nango.log("âœ… hello from action");
  return {
    id: '123',
    status: 'open'
  }
}`;
        const result = zeroYaml.transformAction({
            content: ts,
            action: {
                name: 'testAction',
                type: 'action',
                description: 'Test action',
                version: '1.0.0',
                endpoint: { method: 'POST', path: '/example/github/issues', group: 'Issues' },
                input: 'IssueInput',
                output: ['IssueOutput'],
                scopes: ['repo'],
                usedModels: ['IssueInput', 'IssueOutput']
            },
            models: new Map<string, NangoModel>([
                [
                    'IssueInput',
                    {
                        name: 'IssueInput',
                        fields: [
                            { name: 'title', value: 'string' },
                            { name: 'body', value: 'string' }
                        ]
                    }
                ],
                [
                    'IssueOutput',
                    {
                        name: 'IssueOutput',
                        fields: [
                            { name: 'id', value: 'string' },
                            { name: 'status', value: 'string' }
                        ]
                    }
                ]
            ])
        });
        expect(result).toMatchSnapshot();
    });
});

describe('transformOnEvents', () => {
    it('should transform an on-event script', () => {
        const ts = `import type { NangoSync } from "../../models";

export default async function onEvent(nango: NangoSync): Promise<void> {
  await nango.log("test pre script");
}`;
        const result = zeroYaml.transformOnEvents({
            content: ts,
            eventType: 'pre-connection-deletion'
        });
        expect(result).toMatchSnapshot();
    });
});
