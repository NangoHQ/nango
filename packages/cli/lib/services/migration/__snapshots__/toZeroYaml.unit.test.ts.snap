// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transformAction > should transform an action 1`] = `
"import { z } from "zod";
import { createAction } from "nango";
import type { NangoAction } from "nango";

const IssueInput = z.object({
  title: z.string(),
  body: z.string()
});

const IssueOutput = z.object({
  id: z.string(),
  status: z.string()
});

type IssueInput = z.infer<typeof IssueInput>;
type IssueOutput = z.infer<typeof IssueOutput>;

export default createAction({
  description: "Test action",
  version: "1.0.0",

  endpoint: {
    method: "POST",
    path: "/example/github/issues",
    group: "Issues"
  },

  input: IssueInput,
  output: IssueOutput,
  scopes: ["repo"],

  exec: async (nango, input) => {
    await nango.log("âœ… hello from action");
    return {
      id: '123',
      status: 'open'
    }
  }
});

function foobar(nango: NangoAction) {
   nango.batchSave([{}], 'Model');
}"
`;

exports[`transformOnEvents > should transform an on-event script 1`] = `
"import { createOnEvent } from "nango";

export default createOnEvent({
  event: "pre-connection-deletion",
  description: "pre-connection-deletion event handler",

  exec: async nango => {
    await nango.log("test pre script");
  }
});"
`;

exports[`transformSync > should transform a sync 1`] = `
"import { z } from "zod";
import { createSync } from "nango";
import type { NangoSync } from "nango";

const Model = z.object({
  id: z.string(),
  arrTsType: z.string().array(),
  null: z.nullable(),
  bool: z.boolean(),
  any: z.any(),
  opt: z.number().optional(),
  ref: z.literal("Foo").optional(),
  union: z.union([z.literal("literal1"), z.literal("literal2")]),
  array: z.array(z.union([z.literal("arr1"), z.literal("arr2")])),

  obj: z.object({
    nes: z.literal("ted")
  }),

  dynamicObj: z.object({}).catchall(z.literal("date")),
  union2: z.union([Model, Model.array()]),
  dateOrNull: z.date().optional()
});

const Metadata = z.object({
  foo: z.literal("bar"),
  model: Model
});

const ModelDynamic = z.object({
  id: z.string()
}).catchall(z.string());

type Model = z.infer<typeof Model>;
type Metadata = z.infer<typeof Metadata>;
type ModelDynamic = z.infer<typeof ModelDynamic>;

export default createSync({
  description: "Test sync",
  version: "1.2.3",
  runs: "every hour",
  autoStart: true,
  syncType: "full",
  trackDeletes: false,

  endpoints: [{
    method: "GET",
    path: "top",
    group: "foobar"
  }],

  webhookSubscriptions: ["*"],
  scopes: ["repo", "user"],

  models: {
    Model: Model,
    ModelDynamic: ModelDynamic
  },

  metadata: Metadata,

  exec: async nango => {
      await nango.log('hello');
      await nango.batchSave([{
          'id': 'foobar',
      }], 'Model');

  },

  onWebhook: async (nango, payload) => {
    await nango.log('Received webhook', payload);
    await nango.batchSave([{}], 'Model');
  }
});

function foobar(nango: NangoSync) {
   nango.batchSave([{}], 'Model');
}"
`;
